pkg/webhook/clusterset/validatingWebhook.go

```diff
@@ -0,0 +1,356 @@
+}
+var clusterDeploymentsGVR = metav1.GroupVersionResource{
+	Group:    "hive.openshift.io",
+	Version:  "v1",
+	Resource: "clusterdeployments",
+}
+var clusterPoolsGVR = metav1.GroupVersionResource{
+	Group:    "hive.openshift.io",
+	Version:  "v1",
+	Resource: "clusterpools",
+}
+var clusterClaimsGVR = metav1.GroupVersionResource{
+	Group:    "hive.openshift.io",
+	Version:  "v1",
+	Resource: "clusterclaims",
+}
+
+const (
+	clusterSetLabel = "cluster.open-cluster-management.io/clusterset"
+)
+
+func (a *AdmissionHandler) validateResource(request *v1.AdmissionRequest) *v1.AdmissionResponse {
+	status := &v1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	switch request.Resource {
+	case managedClustersGVR:
+	case clusterDeploymentsGVR:
+	case clusterClaimsGVR:
+	case clusterPoolsGVR:
+	default:
+		return status
+	}
+
+	switch request.Operation {
+	case v1.Create:
+		allowAll, err := a.allowWithoutClusterSetLabel(request.UserInfo)
+		if allowAll {
+			return status
+		} else if err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+				Message: err.Error(),
+			}
+			return status
+		}
+		return a.validateCreateRequest(request)
+	case v1.Update:
+		return a.validateUpdateRequest(request)
+	}
+
+	return status
+}
+
+func (a *AdmissionHandler) validateCreateRequest(request *v1.AdmissionRequest) *v1.AdmissionResponse {
+	status := &v1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	obj := unstructured.Unstructured{}
+	err := obj.UnmarshalJSON(request.Object.Raw)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	switch request.Resource {
+	case clusterDeploymentsGVR:
+		// ignore clusterDeployment created by clusterClaim
+		claimName, _, err := unstructured.NestedString(obj.Object, "spec", "clusterPoolRef", "claimName")
+		if err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+				Message: err.Error(),
+			}
+			return status
+		}
+		if claimName != "" {
+			return status
+		}
+	}
+
+	labels := obj.GetLabels()
+	clusterSetName := ""
+	if len(labels) > 0 {
+		clusterSetName = labels[clusterSetLabel]
+	}
+
+	if clusterSetName == "" {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+			Message: fmt.Sprintf("user %q is forbidden to create a resource without ManagedClusterSet label", request.UserInfo.Username),
+		}
+		return status
+	}
+
+	return a.allowSetClusterSetLabel(request.UserInfo, "", clusterSetName)
+}
+
+func (a *AdmissionHandler) validateUpdateRequest(request *v1.AdmissionRequest) *v1.AdmissionResponse {
```

Comment:
more doc here please.

Reply:
done
// validateUpdateRequest validates:
// 1. allow if the managedClusterSet label is not changed.
// 2. the user must have clusterSet-A and clusterSet-B managedClusterSet/join permission if the managedClusterSet label is update from clusterSet-A to clusterSet-A.
// 3. the user must have clusterSet-A managedClusterSet/join permission if the managedClusterSet label clusterSet-A is removed.